onnxruntime/core/framework/sequential_executor.cc

```diff
@@ -164,9 +164,11 @@ Status SequentialExecutor::Execute(const SessionState& session_state,
                                                      {{"op_name", p_op_kernel->KernelDef().OpName()}});
     }
 
-    // free ml-values corresponding to this node
-    VLOGS(logger, 1) << "Releasing node ML values after computing kernel: " << p_op_kernel->Node().Name();
-    ORT_RETURN_IF_ERROR(ReleaseNodeMLValues(frame, seq_exec_plan, node_exec_plan, logger));
+    if (p_op_kernel->Node().GetExecutionProviderType() != kCudaExecutionProvider || queue_id == 0) {
+      // free ml-values corresponding to this node
+      VLOGS(logger, 1) << "Releasing node ML values after computing kernel: " << p_op_kernel->Node().Name();
+      ORT_RETURN_IF_ERROR(ReleaseNodeMLValues(frame, seq_exec_plan, node_exec_plan, logger));
```

Comment:
>ReleaseNodeMLValues [](start = 26, length = 19)

Seems this would never be called if the node runs in queue_id != 0 in CUDA, which may cause some other models to fail?

Reply:
Maybe a better fix is to change [https://github.com/Microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cuda/cuda_execution_provider.cc#L18-L38](https://github.com/Microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cuda/cuda_execution_provider.cc#L18-L38), to let CUDA memcpy happen on default stream. That would slow down GPU a little bit, but should not have major performance or regression risk.

---
In reply to: [276409451](https://github.com/Microsoft/onnxruntime/pull/812#discussion_r276409451) [](ancestors = 276409451)

Reply:
Your understanding is correct. This PR aims to disable the clean-up of `MLValue` when the corresponding `queue_id != 0` in CUDA (the `queue_id` is the CUDA stream id in this case). If you take a look at [the source code of CUDA provider](https://github.com/Microsoft/onnxruntime/blob/2a2de42bb257c7288103dd96ececebecedfb6c57/onnxruntime/core/providers/cuda/cuda_execution_provider.h#L20), you can find that the `queue_id` will only be non-zero for `kCudaStreamCopyIn` and `kCudaStreamCopyOut` stream type, which are exclusively used in host-device memcpy operator nodes. For the `MLValue`s in these nodes, `MLValue`s are supposed to be cleaned up at the end of one execution (at least I supposed so, because I saw in the [parallel executor](https://github.com/Microsoft/onnxruntime/blob/master/onnxruntime/core/framework/parallel_executor.cc), MLValue is never explicitly released).

Nevertheless, I'm not 100% sure if it could break the models in your internal model collection. Maybe you could just give it a try? And let's see if everything is going well?

Reply:
> Maybe a better fix is to change https://github.com/Microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cuda/cuda_execution_provider.cc#L18-L38, to let CUDA memcpy async happen on default stream. That would slow down GPU a little bit, but should not have major performance or regression risk.

This solution makes sense because the expectation of a `SequentialExector` is to execute all nodes sequentially, i.e., in a single stream. If could be better if we can determine the CUDA stream during runtime (depend on whether it's in `SequentialExecutor` or `ParallelExecutor`). That might need more change in the framework.

Let me change the default cuda stream id to default stream for now. For separated stream for memcpy, you may want to do it in another PR.

Reply:
We have identified a bug in the current fence system which will result in the mem override issue. This has been seen from another model that Du worked on. The real fix should be place fence before MLValue get released.

Reply:
@ybrnathan Let's hope the bug is the same as this one. But I still have concern as in https://github.com/Microsoft/onnxruntime/pull/812#issuecomment-481914326 that this one may not relate to the fence system.